qemu/arm: Extend CI testing to include ArmV7-M hardfp targets.

agatti opened the pull request:
### Summary

As a followup to #17737, this PR introduces a QEMU Arm target that has a hardware floating point unit (`MPS2_AN500`), and then lets CI use it for selected operations.

`MPS2_AN500` is a Cortex-M7 with a single- and double-precision floating point unit, but it is otherwise equivalent to the `MPS2_AN385` target for MicroPython's needs.

To further validate that the FPU is used and initialised correctly, the machine exception handler for ArmV7-M targets was extended to dump the full set of debug registers.  If the FPU is used incorrectly (as in, access to the appropriate machine coprocessor was not enabled beforehand) then the `CFSR.NOCP` bit should appear to be set in the exception handler dump.

In order to correctly test the hardfp natmods we need to have a validated hardfp interpreter, so we must have a full `MPS2_AN500` build and test cycle as well.  To reduce the impact on CI times I've split the original CI step into two discrete units, as in "ci_qemu_build_arm_thumb_softfp" and "ci_qemu_build_arm_thumb_hardfp", so they can run in parallel whenever possible.

Manual building and testing of the QEMU port isn't otherwise affected, the default Arm machine is still `MPS2_AN385` and it will retain the same floating point configuration as before (`MICROPY_FLOAT_IMPL_FLOAT`).

Concidentally, I've made QEMU's Makefile a bit more flexible when it comes to floating point settings.  I've replicated the setup from the `stm32` port, so the same flag names apply here as well.

### Testing

Besides manual testing, CI should (hopefully) run the relevant step without errors; tests passed on my branch.

### Trade-offs and Alternatives

There will be a (hopefully) minor increase in CI build times.  However, having two separate targets is necessary to catch issues with FPU opcodes being used in softfp contexts for example.  Making the hardfp variant the default would still call for a softfp variant to be there anyway.

Things could still be improved though, as in ArmV6-M builds could be tested in an emulated Cortex-M0+ machine to make sure no ArmV7-M opcodes are used by GCC or by viper/native emitters.

agatti committed:
qemu/mcu/arm/errorhandler: Add ARMv7-M debug registers.

This commit extends the QEMU port's CPU error handler for the Arm target
by printing out in detail the ARMv7-M debug registers if the firmware is
compiled for such a target.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tools/ci.sh: Extend Arm testing to include hardfp targets.

This commit lets CI extend the testing scope of the QEMU Arm target, by
letting it perform the usual battery of tests (interpreter and natmods)
also on hardfp targets.

The default board for Arm testing lacks hardware floating point support,
so natmods weren't tested in that specific configuration.  With the
introduction of the "MPS_AN500" QEMU target, now this is made possible
as said board emulates a Cortex-M7 machine with a single- and
double-precision floating point unit.

To reduce the impact on build times, the "ci_qemu_build_arm_thumb" CI
step was split in two: "ci_qemu_build_arm_thumb_softfp" and
"ci_qemu_build_arm_thumb_hardfp" - so hopefully those can run in
parallel whenever possible.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
qemu/Makefile: Allow overriding floating point mode by boards.

This commit lets board use a different floating point mode rather than
the usual soft-float that was the original default for all QEMU-based
boards.

The configuration options are the same available in the "stm32" port.
Boards can set "MICROPY_FLOAT_IMPL" to either "float", "double", or
"none" to indicate which floating point mode they want, and optionally
"SUPPORTS_HARDWARE_FP_SINGLE" or "SUPPORTS_HARDWARE_FP_DOUBLE" can be
set to 1 to further indicate the hardware capabilities of the hardware
floating point unit, if present.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
qemu/arm: Add definition for the MPS2_AN500 machine.

This commit introduces a new target for the QEMU port called
"MPS2_AN500", an ARMv7-M machine with a Cortex-M7 CPU and
single-/double-precision floating point unit.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
qemu/mcu/arm/mps2.ld: Add .ARM.exidx section to the linkerscript.

This commit fixes a linking issue on certain Arm toolchains where
library code is compiled with exception support.

If a library with exception support is included in the MicroPython
build, the linker had no place to put the stack unwinding tables
necessary to perform exception handling at runtime.  This change adds a
new section to the linkerscript (and therefore the final ELF file) where
that data can be placed into.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge commented:
> For the record, my `arm-none-eabi-binutils` package is version 2.43-1, and `arm-none-eabi-newlib` is 4.5.0.20241231-1.

I have exactly the same versions.

And I just found that a fresh checkout of this PR -- WITHOUT REBASING -- builds fine.  It's after a rebase on latest master that it doesn't link.  I suspect you'll also see the problem after rebasing (and it's maybe because of changes to floating point constant folding, bringing in some additional libm code?).

Applying the above exidx patch gets it linking after a rebase.

dpgeorge commented:
Note that `tests/run-tests.py` has the following:
```
    "qemu": (
        # Skip tests that require Cortex-M4.
        "inlineasm/thumb/asmfpaddsub.py",
        "inlineasm/thumb/asmfpcmp.py",
        "inlineasm/thumb/asmfpldrstr.py",
        "inlineasm/thumb/asmfpmuldiv.py",
        "inlineasm/thumb/asmfpsqrt.py",
    ),
```
and so these tests are skipped on the MPS2_AN500, even though they could run and pass on it.  Not sure if that's worth improving in this PR, or for a follow-up PR (you can decide).

agatti committed:
tests/run-tests.py: Enable Arm inlineasm FPU tests if possible.

This commits lifts the unconditional restriction on inline assembler FPU
tests for the Qemu platform, and makes said restriction conditional to
the lack of an available floating point unit on the running platform.

The Qemu platform supported only emulated machines that could target up
to a Cortex-M3, so an ArmV7-M target that had no support for floating
point.  With the addition of MPS2_AN500 to the list of emulated targets
the range was extended to cover up to Cortex-M7, so a floating point
unit may possibly be available and thus able to run the FPU inlineasm
tests.

For that, the test runner was changed to detect the running architecture
when checking the target capabilities; if the target reports its
MicroPython architecture to be either "armv7emsp" or "armv7emdp"
(providing single-precision and double-precision floating point unit
support respectively) then the FPU-only inline tests are not put into
the blocked tests list.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
> And I just found that a fresh checkout of this PR -- WITHOUT REBASING -- builds fine. It's after a rebase on latest master that it doesn't link. I suspect you'll also see the problem after rebasing (and it's maybe because of changes to floating point constant folding, bringing in some additional libm code?).
> 
> Applying the above exidx patch gets it linking after a rebase.

Indeed I was able to reproduce (and fix) the issue after a rebase - thanks for the hint.  I've added a commit to the PR with the linkerscript changes addressing this issue.

> Note that `tests/run-tests.py` has the following:
> [...]
> and so these tests are skipped on the MPS2_AN500, even though they could run and pass on it. Not sure if that's worth improving in this PR, or for a follow-up PR (you can decide).

No worries, I've just updated the test runner to put those tests into the block list if the reported arch is neither `armv7emsp` nor `armv7emdp`.  I've tested it with `MPS2_AN385`, `MPS2_AN500`, and `SABRELITE` and the set of tests being executed matches what they are supposed to run.

dpgeorge reviewed (approved):
Looks good now thanks!

Tested locally on qemu/MPS2_AN500, and stm32/PYBV11.  Tests pass.
